juju/juju
# [JUJU-3763] Add metrics to the changestream and dbaccessor packages

This patch adds new metrics for the changestream (both stream and eventmultiplexer included) package, for monitoring current number of subscriptions, errors, number of dispatched changes and watermarks. 
Also, on the dbaccessor package a new metric was added for counting the number of transaction retries, keeping a separate metric for total number of transactions.

Many of the changes are simply fixing existing tests to account for the new call to observe metrics.

## Checklist

*If an item is not applicable, use `~strikethrough~`.*


## QA steps

```
go test github.com/juju/juju/worker/changestream/... -gocheck.v 
go test github.com/juju/juju/worker/dbaccessor/... -gocheck.v 
```
Also, bootstrap and deploy anything (e.g. ubuntu) then ssh to the controller and retrieve the metrics:

```
juju bootstrap localhost c
juju add-model m
juju deploy ubuntu
juju ssh -m controller 0
ubuntu@juju-509cd2-0:~$ juju_metrics
```
You should have a big list of prometheus metrics, but make sure you see (or grep) some of the new metrics, for example:

```
# HELP juju_db_watermark_inserts_total Total number of watermark insertions on the changelog witness table.
# TYPE juju_db_watermark_inserts_total counter
juju_db_watermark_inserts_total 355
# HELP juju_db_watermark_retries_total Total number of watermark retries on the changelog witness table.
# TYPE juju_db_watermark_retries_total counter
juju_db_watermark_retries_total 355
```
or 
```
# HELP juju_db_subscription_count The total number of subscriptions.
# TYPE juju_db_subscription_count gauge
juju_db_subscription_count 1
```
